plugins/cody-chat/src/com/sourcegraph/cody/chat/access/TokenStorage.java

```diff
@@ -1,33 +1,142 @@
 package com.sourcegraph.cody.chat.access;
 
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences.INodeChangeListener;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences.NodeChangeEvent;
+import org.eclipse.core.runtime.preferences.InstanceScope;
 import org.eclipse.e4.core.di.annotations.Creatable;
+import org.eclipse.equinox.security.storage.EncodingUtils;
 import org.eclipse.equinox.security.storage.ISecurePreferences;
 import org.eclipse.equinox.security.storage.SecurePreferencesFactory;
 import org.eclipse.equinox.security.storage.StorageException;
-
-import java.io.IOException;
+import org.osgi.service.prefs.BackingStoreException;
 
 import jakarta.inject.Singleton;
 
 @Creatable
 @Singleton
 public class TokenStorage {
-	
-	private static String NODE_ID = "/com/sourcegraph/cody/chat";
-	private static String KEY = "token";
-	
-	private ISecurePreferences preferences = SecurePreferencesFactory.getDefault().node(NODE_ID);
-	
-	public void put(String token) throws StorageException, IOException {
-		preferences.put(KEY, token, true);
-		preferences.flush();
+
+	public static class Profile {
+		public final String name;
+		public final String url;
+		public Profile(String name, String url) {
+			this.name = name;
+			this.url = url;
+		}
+	}
+
+	private static String SECURE_NODE_ID = "/com/sourcegraph/cody/chat/tokens";
+	private static String SETTING_NODE_ID = "cody-chat/tokens"; // Path for settings node should start with a plugin ID
+	private static String PROFILES_NODE_ID = SETTING_NODE_ID + "/profiles";
+
+	private ISecurePreferences secureStorage = SecurePreferencesFactory.getDefault().node(SECURE_NODE_ID);
+	private IEclipsePreferences settingsStorage = InstanceScope.INSTANCE.getNode(SETTING_NODE_ID);
+	private IEclipsePreferences profileStorage = InstanceScope.INSTANCE.getNode(PROFILES_NODE_ID);
+	// Profiles are associated with a workspace. To associate them with Eclipse
+	// installation change `InstanceScope` to `ConfigurationScope`.
```

Comment:
Would be good to mention parallels to VSC/JB

* VSC: User / Workspace
* JB: Application / Project
* Eclipse: Instance / Configuration